cargo test segfaulting on i386

Debianhas recently upgraded rust-bindgen to 0.55.1 in unstable. To keep dependencies satisfiable the update of bindgen required updating rust-onig-sys as well and it was upgraded to 69.5.1. 

However, after the update of bindgen onig-sys it was discovered that the tests for onig started failing on i386 and armhf and this is currently blocking the testing migration of the new rust-bindgen package. On i386 the tests fail with a segfault and on armhf they faile with a bus error. I performed further testing on i386.

First I tried updating to the latest version of the onig crate and the error persisted.

I then tried doing a test outside of the "debcargo" environment, I first tried using
an upstream git checkout but that failed to find it's embedded copy of libonig
so I decided to use cargo clone instead.

cargo install cargo clone
cargo clone onig
cd onig
cargo test

And again that failed in the same way

    running 42 tests
    test buffers::tests::byte_buffer_create ... ok
    test buffers::tests::rust_bytes_encoding_is_ascii ... ok
    test buffers::tests::rust_string_ptr_offsets_are_valid ... ok
    test buffers::tests::rust_bytes_ptr_offsets_are_valid ... ok
    test buffers::tests::rust_string_encoding_is_utf8 ... ok
    test find::tests::test_captures_iter ... ok
    test find::tests::test_find_iter ... ok
    test find::tests::test_find_iter_empty_after_match ... ok
    test find::tests::test_find_iter_one_zero_length ... ok
    test find::tests::test_regex_captures ... ok
    test find::tests::test_regex_subcaptures ... ok
    test find::tests::test_regex_subcapturespos ... ok
    test match_param::test::create_default_match_param ... ok
    test find::tests::test_zero_length_matches_jumps_past_match_location ... ok
    test match_param::test::set_max_stack_size_limit ... ok
    test match_param::test::set_retry_limit_in_match ... ok
    test find::tests::test_find_iter_many_zero_length ... ok
    error: test failed, to rerun pass '--lib'
    
    Caused by:
      process didn't exit successfully: `/onig/target/debug/deps/onig-badb9b5fa42b5250` (signal: 11, SIGSEGV: invalid memory reference)

I suspect this has something to do with either our bindings to `libonig` not having the right
layouts for some things in the headers, or the way we're compiling `libonig` not quite being
correct on 32 bit systems.

On a 32 bit Ubuntu VM I'm seeing the following when trying to compile:

```
   Compiling onig_sys v69.5.1 (/vagrant/onig_sys)
warning: oniguruma/src/regparse.c: In function ‘onig_st_lookup_strend’:
warning: oniguruma/src/regparse.c:528:32: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:    return onig_st_lookup(table, (st_data_t )(&key), value);
warning:                                 ^
warning: oniguruma/src/regparse.c: In function ‘onig_st_insert_strend’:
warning: oniguruma/src/regparse.c:543:34: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:    result = onig_st_insert(table, (st_data_t )key, value);
warning:                                   ^
warning: oniguruma/src/regparse.c: In function ‘onig_st_lookup_callout_name_table’:
warning: oniguruma/src/regparse.c:624:32: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:    return onig_st_lookup(table, (st_data_t )(&key), value);
warning:                                 ^
warning: oniguruma/src/regparse.c: In function ‘st_insert_callout_name_table’:
warning: oniguruma/src/regparse.c:644:34: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:    result = onig_st_insert(table, (st_data_t )key, value);
warning:                                   ^
warning: oniguruma/src/regparse.c: In function ‘onig_foreach_name’:
warning: oniguruma/src/regparse.c:797:33: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:      onig_st_foreach(t, i_names, (HashDataType )&narg);
warning:                                  ^
warning: oniguruma/src/regparse.c: In function ‘onig_renumber_name_table’:
warning: oniguruma/src/regparse.c:825:41: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:      onig_st_foreach(t, i_renumber_name, (HashDataType )map);
warning:                                          ^
warning: oniguruma/src/regparse.c: In function ‘name_add’:
warning: oniguruma/src/regparse.c:1007:31: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:                                (HashDataType )e);
warning:                                ^
warning: oniguruma/src/regparse.c: In function ‘callout_name_entry’:
warning: oniguruma/src/regparse.c:1445:38: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:                                       (HashDataType )e);
warning:                                       ^
warning: oniguruma/src/regparse.c: In function ‘setup_ext_callout_list_values’:
warning: oniguruma/src/regparse.c:1807:21: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:                      (st_data_t )ext);
warning:                      ^
warning: oniguruma/src/unicode.c: In function ‘onig_unicode_define_user_property’:
warning: oniguruma/src/unicode.c:1110:29: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
warning:                              (hash_data_type )((void* )e));
warning:                              ^
   Compiling onig v6.1.0 (/vagrant/onig)
```

Doesn't _sound_ good.

--- a/.gitignore
+++ b/.gitignore
@@ -4,3 +4,4 @@ target/
 .vagrant/
 *.orig
 .idea/
+*.log

--- a/Vagrantfile
+++ b/Vagrantfile
@@ -2,10 +2,11 @@
 # vi: set ft=ruby :
 
 Vagrant.configure(2) do |config|
-  config.vm.box = "ubuntu/xenial64"
-  config.vm.provision "shell", inline: <<-SHELL
+  # config.vm.box = "ubuntu/xenial64"
+  config.vm.box = "ubuntu/xenial32"
+  config.vm.provision "shell", privileged: false, inline: <<-SHELL
     sudo apt-get update
-    sudo apt-get install -y clang make binutils cmake
-    curl -sSf https://static.rust-lang.org/rustup.sh | sh -s -- -y 2>/dev/null
+    sudo apt-get install -y clang-8 make binutils cmake
+    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y 2>/dev/null
   SHELL
 end

--- a/onig_sys/build.rs
+++ b/onig_sys/build.rs
@@ -113,7 +113,9 @@ fn compile() {
             #define HAVE_INTTYPES_H 1
             #define SIZEOF_INT 4
             #define SIZEOF_SHORT 2
-            #define SIZEOF_LONG {}
+            #define SIZEOF_LONG {0}
+            #define SIZEOF_VOIDP {0}
+            #define SIZEOF_LONG_LONG 8
         ",
                 if bits == "64" { "8" } else { "4" }
             ),

